clerk/verification.go

```diff
@@ -0,0 +1,72 @@
+package clerk
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+const (
+	CookieSession       = "__session"
+	QueryParamSessionId = "_clerk_session_id"
+	OriginHeader        = "Origin"
+)
+
+type VerificationService service
+
+type verifyRequest struct {
+	Token string `json:"token"`
+}
+
+func (s *VerificationService) Verify(req *http.Request) (*Session, error) {
+	if req == nil {
+		return nil, errors.New("cannot verify empty request")
+	}
+	cookie, err := req.Cookie(CookieSession)
+	if err != nil {
+		return nil, errors.New("couldn't find cookie " + CookieSession)
+	}
+
+	sessionToken := cookie.Value
+	sessionId := req.URL.Query().Get(QueryParamSessionId)
+
+	if !isXHR(req) && req.Method == "GET" && sessionId == "" {
+		return s.useClientActiveSession(sessionToken)
+	}
+
+	if sessionId == "" {
+		return nil, errors.New(fmt.Sprintf("no session id is specified via the %s query parameter", QueryParamSessionId))
+	}
+
+	return s.client.Sessions().Verify(sessionId, sessionToken)
+}
+
+func isXHR(req *http.Request) bool {
+	headers := req.Header
```

Comment:
💡 For being more future-proof, we could also check here that `req` is not nil. I know that currently only `Verify` calls into this and this is redundant, but that could change in the future.

Reply:
I'm not a huge fan of null-checks and error checks everywhere. 

I believe that we should be defensive and make the necessary checks on the boundaries of our application (where we're dealing with external inputs) but other than that, trust everything else that is in our control.